src/ProductConstructionService/ProductConstructionService.Deployment/Deployer.cs

```diff
@@ -222,33 +237,81 @@ private string GetLogsLink(string revisionName)
 
     private async Task<ProcessExecutionResult> InvokeAzCLI(string[] command)
     {
+        string[] fullCommand = [.. command, .. DefaultAzCliParameters];
+        _logger.LogInformation($"Invoking az cli command `{command}`", string.Join(' ', fullCommand));
         return await _processManager.Execute(
             Path.GetFileName(_options.AzCliPath),
-            [
-                .. command,
-                .. DefaultAzCliParameters
-            ],
+            fullCommand,
             workingDir: Path.GetDirectoryName(_options.AzCliPath));
     }
 
-    private async Task StopProcessingNewJobs()
+    private async Task StopProcessingNewJobs(string activeRevisionName)
     {
-        Console.WriteLine("Stopping the service from processing new jobs");
-        await _pcsClient.Status.StopPcsWorkItemProcessorAsync();
+        _logger.LogInformation("Stopping the service from processing new jobs");
 
-        string status;
+        var replicas = await GetRevisionReplicaStates(activeRevisionName);
         try
         {
+            foreach (var replica in replicas)
+            {
+                await replica.FinishWorkItemAndStopAsync();
+            }
+
+            bool allReplicasStopped;
             do
             {
-                status = await _pcsClient.Status.GetPcsWorkItemProcessorStatusAsync();
+                allReplicasStopped = true;
+                foreach (var replica in replicas)
+                {
+                    var state = await replica.GetStateAsync();
+                    _logger.LogInformation("Replica {replicaName} has status: {state}",
+                        replica.ReplicaName,
+                        state);
+                    if (state != WorkItemProcessorState.Stopped)
+                    {
+                        allReplicasStopped = false;
+                    }
+                }
```

Comment:
We could do without the boolean and also paralelize this in one go if we start all the tasks at once with
`replicas.Select(replica => replica.GetStateAsync())` and then await all and do `Any(state => state != WorkItemProcessorState.Stopped)`?

We could then just `break` after the `Any()` and log+sleep at the end of the loop and we wouldn't need the utility even.